src/scores/continuous/threshold_weighted_impl.py

```diff
@@ -0,0 +1,687 @@
+
+    Returns:
+        array of function values for the function g_j.
+
+    Note:
+        Requires a < b. This is tested in :py:func:`_auxiliary_funcs`.
+    """
+    # results correspond to each case in the first row of Table B1, Taggart (2022).
+    result1 = 0
+    result2 = x - a
+    result3 = b - a
+
+    result = result2.where(x < b, result3).where(x >= a, result1)
+    result = result.where(~np.isnan(x), np.nan)
+
+    return result
+
+
+def _phi_j_rect(a: EndpointType, b: EndpointType, x: xr.DataArray) -> xr.DataArray:
+    """
+    Returns values of a convex function phi_j, where phi_j is obtained by integrating
+    a rectangular weight function. The weight is 1 on the interval [a, b) and 0 otherwise.
+    The formula is based on the second row of Table B1 from Taggart (2022).
+
+    Args:
+        a: left endpoint of interval where weight = 1. Can be `-np.inf`.
+        b: right endpoint of the interval where weight = 1. Can be `np.inf`.
+        x: points where phi_j is to be evaluated.
+
+    Returns:
+        array of function values for the function phi_j.
+
+    Note:
+        Requires a < b. This is tested in :py:func:`_auxiliary_funcs`.
+    """
+    # results correspond to each case in the second row of Table B1, Taggart (2022).
+    result1 = 0
+    result2 = 2 * (x - a) ** 2
+    result3 = 4 * (b - a) * x + 2 * (a**2 - b**2)
+
+    result = result2.where(x < b, result3).where(x >= a, result1)
+    result = result.where(~np.isnan(x), np.nan)
+
+    return result
+
+
+def _phi_j_prime_rect(a: EndpointType, b: EndpointType, x: xr.DataArray) -> xr.DataArray:
+    """
+    The subderivative of `_phi_j_rect(a, b, x)` w.r.t. x.
+    """
+    return 4 * _g_j_rect(a, b, x)
+
+
+def _g_j_trap(a: EndpointType, b: EndpointType, c: EndpointType, d: EndpointType, x: xr.DataArray) -> xr.DataArray:
+    """
+    Returns values of a nondecreasing function g_j, where g_j is obtained by integrating
+    a trapezoidal weight function. The weight is 1 on the interval (b, c) and 0 outside
+    the interval (a,d). The formula is based on the third row of Table B1 from Taggart (2022).
+
+    Args:
+        a: left endpoint of interval where weight > 0.
+        b: left endpoint of the interval where weight = 1.
+        c: right endpoint of the interval where weight = 1.
+        d: right endpoint of interval where weight > 0.
+        x: points where g_j is to be evaluated.
+
+    Returns:
+        array of function values for the function g_j.
+
+    Note:
+        Requires a < b < c < d. This is tested in :py:func:`_auxiliary_funcs`.
+    """
+    # results correspond to each case in the third row of Table B1, Taggart (2022).
+    result0 = 0
+    result1 = (x - a) ** 2 / (2 * (b - a))
+    result2 = x - (b + a) / 2
+    result3 = -((d - x) ** 2) / (2 * (d - c)) + (d + c - a - b) / 2
+    result4 = (d + c - a - b) / 2
+    result = (
+        result1.where(x >= a, result0)
+        .where(x < b, result2)
+        .where(x < c, result3)
```

Comment:
Just a suggestion, easier to debug by splitting this out into different lines rather than using chaining.

Reply:
Removed chaining